File: app/javascript/channels/game_channel.js
import consumer from "./consumer"
import shakeEffect from "../packs/shake_effect";
import {updatePlayersList} from "../packs/game/targetQuestion";
import questionTemplate from "../packs/templates/question";

$( document ).on('turbolinks:load', function() {
  const gameContainer = $('#game-container')
  if (gameContainer.length <= 0) { return }

  const gameId = gameContainer.data('game-id')
  const playerId = gameContainer.data('player-id')
  const startBtn = $('#start-game-btn')
  const nextQuestionBtn = $('#next-question-btn')
  const questionBody = $('#question_body')

  const params = {
    channel: 'GameChannel',
    game_id: gameId,
    player_id: playerId,
  }

  consumer.subscriptions.create(params, {
    // called once from very beginning
    initialized() {

    },

    // called on subscribing to channel
    connected() {
      console.log('connected')
      this.setup()
    },

    // called on disconnecting
    disconnected() {
      console.log('disconnected')
    },

    // called when server publish something
    received(data) {
      console.log('received: ', data)
      switch (data.type) {
        case 'game_state':
          if (data.message === 'started') { document.location.reload() }
          break
        case 'question':
          this.renderQuestion(data.message)
          break
      }
      updatePlayersList(data.players, playerId)
    },

    // custom instance methods
    setup() {
      this.addListeners()
    },

    startGame() {
      this.perform("start_game")
    },

    nextQuestion() {
      this.perform("next_question")
    },

    renderQuestion(data) {
      questionBody.html(questionTemplate(data))
      shakeEffect(questionBody[0])
      if (data.question == null) { nextQuestionBtn.hide() } else { nextQuestionBtn.show() }
    },

    addListeners() {
      startBtn.click(e => {
        this.startGame()
        document.location.reload();
      })

      nextQuestionBtn.click(e => {
        e.stopPropagation()
        this.nextQuestion()
      })
    },
  });
})

File: app/channels/game_channel.rb
class GameChannel < ApplicationCable::Channel
  attr_reader :game, :player, :channel_name

  # called on subscribing, work's as initialize for GameChannel instance
  def subscribed
    @player = Player.find params[:player_id]
    @game = Game.find params[:game_id]
    @channel_name = "Game-#{params[:game_id]}"

    stream_from(@channel_name)

    broadcast_question if @game.active?
  end

  # called on disconnecting
  def unsubscribed; end

  # custom public method
  def start_game
    game.active!
    broadcast_game_started
  end

  # custom public method
  def next_question
    Rails.cache.delete(channel_name)
    broadcast_question
  end

  private

  def current_question
    Rails.cache.fetch(channel_name, expires_in: 1.hours) do
      QuestionsManager.new(game).next_question
    end
  end

  def broadcast_game_started
    broadcast(:game_state, :started)
  end

  def broadcast_question
    broadcast(:question, current_question)
  end

  def broadcast(type, message)
    data = { type: type, message: message, players: all_players_hash }
    # broadcasting data to each subscriber
    ActionCable.server.broadcast(channel_name, data)
  end

  def all_players_hash
    game.players.map { |p| { id: p.id, name: p.name} }
  end
end
